main.go

```diff
@@ -12,6 +24,125 @@ func init() {
 	_ = flag.Set("logtostderr", "true")
 }
 
+func GetCurrentNode(clientset kubernetes.Clientset) (*corev1.Node, error) {
+	nodeName := os.Getenv("NODE_NAME")
+	if nodeName == "" {
+		return nil, errors.New("Environment variable NODE_NAME is not set")
+	}
+
+	node, err := clientset.CoreV1().Nodes().Get(context.TODO(), nodeName, metav1.GetOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("failed to get node: %w", err)
+	}
+
+	return node, nil
+}
+
+func UpdateNodeLabels(
+	clientset *kubernetes.Clientset,
+	instanceData *metadata.InstanceData,
+) error {
+	if instanceData == nil {
+		return fmt.Errorf("instance data received from Linode metadata service is nil")
+	}
+
+	node, err := GetCurrentNode(*clientset)
+	if err != nil {
+		return fmt.Errorf("failed to get the node: %w", err)
+	}
+
+	klog.Infof("Updating node labels with Linode instance data: %v", instanceData)
+
+	node.Labels["linode_label"] = instanceData.Label
+	node.Labels["linode_id"] = strconv.Itoa(instanceData.ID)
+	node.Labels["linode_region"] = instanceData.Region
+	node.Labels["linode_type"] = instanceData.Type
+	node.Labels["linode_host"] = instanceData.HostUUID
+
+	_, err = clientset.CoreV1().Nodes().Update(context.TODO(), node, metav1.UpdateOptions{})
+
+	if err != nil {
+		klog.Errorf("Failed to update labels: %s", err.Error())
+		return err
+	}
+
+	return nil
+}
+
+func GetClientset() (*kubernetes.Clientset, error) {
+	config, err := rest.InClusterConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	clientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return nil, err
+	}
+
+	return clientset, nil
+}
+
+func StartDecorator(client metadata.Client, clientset *kubernetes.Clientset, interval time.Duration) {
+	instanceData, err := client.GetInstance(context.Background())
+	if err != nil {
+		klog.Errorf("Failed to get the initial instance data: %s", err.Error())
+	} else {
+		err = UpdateNodeLabels(clientset, instanceData)
+	}
+
+	if err != nil {
+		klog.Error(err)
+	}
+
+	instanceWatcher := client.NewInstanceWatcher(
+		metadata.WatcherWithInterval(interval),
+	)
+
+	go instanceWatcher.Start(context.Background())
+
+	for {
+		select {
+		case data := <-instanceWatcher.Updates:
+			err = UpdateNodeLabels(clientset, data)
+			if err != nil {
+				klog.Error(err)
```

Comment:
this could lead to a situation where the labels are never updated on the node yet the decorator continues to happily run, e.g. if the kube api errors out and we only log the error we will never receive the Update again from the watcher... some ideas...
1) you could leave this to a best effort... maybe some sort of back off and retry for now?
2) error out all the way and force the pod to restart which would start the update process on boot which does a fresh pull of metadata... 
3) maybe do both? have any better ideas?

Reply:
I think just `klog .Fatal` would be fine. I am a little bit worried about infinity restart loop, but I think k8s may have some limit to handle it? like number of restarts allowed during a timeframe?

Reply:
ya that's a crash loop back off which is fine for kube to just handle. the scheduler will start an exponential backoff after a bit and you wont even notice it

Reply:
I just made it to be `klog.Fatal`. Consider this as solved. Thank you for the suggestion!